logproof/src/bfv_statement.rs

```diff
@@ -0,0 +1,861 @@
+        statements: Vec<BfvProofStatement<Ciphertext, PublicKey>>,
+        messages: Vec<Plaintext>,
+        witness: Vec<BfvWitness<SecretKey>>,
+    }
+
+    struct BFVTestContext {
+        ctx: Context,
+        params: EncryptionParameters,
+        public_key: PublicKey,
+        encryptor: Encryptor,
+    }
+
+    impl BFVTestContext {
+        fn new() -> Self {
+            let plain_modulus = PlainModulus::raw(1153).unwrap();
+            let coeff_modulus =
+                CoefficientModulus::bfv_default(1024, SecurityLevel::TC128).unwrap();
+            let params = BfvEncryptionParametersBuilder::new()
+                .set_poly_modulus_degree(1024)
+                .set_coefficient_modulus(coeff_modulus)
+                .set_plain_modulus(plain_modulus)
+                .build()
+                .unwrap();
+            let ctx = Context::new(&params, false, SecurityLevel::TC128).unwrap();
+            let gen = KeyGenerator::new(&ctx).unwrap();
+            let public_key = gen.create_public_key();
+            let secret_key = gen.secret_key();
+            let encryptor =
+                Encryptor::with_public_and_secret_key(&ctx, &public_key, &secret_key).unwrap();
+
+            BFVTestContext {
+                ctx,
+                params,
+                public_key,
+                encryptor,
+            }
+        }
+
+        // If we ever speed up SDLP, this would be nice to run in a loop. As is, it's too slow.
+        #[allow(unused)]
+        fn random_fixture(&self) -> TestFixture {
+            let mut rng = rand::thread_rng();
+            let num_statements = rng.gen_range(1..=3);
+            let num_duplicate_msgs = rng.gen_range(0..num_statements);
+            self.random_fixture_with(num_statements, num_duplicate_msgs)
+        }
+
+        /// Note that `num_duplicate_msgs` is in _addition_ to the first message about a given
+        /// index. So [0, 0, 0] contains `num_duplicate_msgs == 2`.
+        //
+        // TODO add private statements once we have a way to generate them, accepting
+        // num_public_statements, num_private_statements.
+        fn random_fixture_with(
+            &self,
+            num_statements: usize,
+            num_duplicate_msgs: usize,
+        ) -> TestFixture {
+            assert!(num_duplicate_msgs < num_statements);
+            let num_msgs = num_statements - num_duplicate_msgs;
+
+            // all the messages
+            let messages = (0..num_msgs)
+                .map(|_| self.random_plaintext())
+                .collect::<Vec<_>>();
+
+            let mut statements = Vec::with_capacity(num_statements);
+            let mut witness = Vec::with_capacity(num_statements);
+
+            // statements without duplicate messages
+            for (i, msg) in messages.iter().enumerate() {
+                let (ct, u, e, r) = self.encryptor.encrypt_return_components(msg).unwrap();
+                statements.push(BfvProofStatement::PublicKeyEncryption {
```

Comment:
Do we have any tests with secret encryptions?

Reply:
No.. as discussed, symmetric encryptions are on hold until after they're exposed from SEAL. Hence the `todo!()` above.